### PR TITLE
Returns grpc status corresponding to the error

### PR DESCRIPTION
**Pull request by hlts2**
Signed-off-by: hlts2 <hiroto.funakoshi.hiroto@gmail.com>

### Description:

Currently, it returns an unkow error without checking the error in detail.
Add an implementation that handles the error in detail and returns the appropriate grpc status.

e.g)

cassandra
```
Unavailable error ...etc
```

### Environment:

- Golang Version: 1.13.5
- Docker Version: 19.03.5
- Kubernetes Version: 1.16.3
- NGT Version: 1.8.4

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensure all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/meta/cassandra/handler/grpc/handler.go [hunk 1]
```diff
@@ -48,10 +48,13 @@ func (s *server) GetMeta(ctx context.Context, key *payload.Meta_Key) (*payload.M
...
 			log.Warnf("[GetMeta]\tnot found\t%v\t%+v", key.GetKey(), err)
-			return nil, status.WrapWithNotFound(fmt.Sprintf("GetMeta API key %s not found", key.GetKey()), err, info.Get())
+			return nil, status.WrapWithNotFound(fmt.Sprintf("GetMeta API Cassandra key %s not found", key.GetKey()), err, info.Get())
+		} else if errors.IsErrCassandraUnavailable(errors.UnWrapAll(err)) {
+			log.Warnf("[GetMeta]\tunavailable\t%+v", err)
+			return nil, status.WrapWithUnavailable(fmt.Sprintf("GetMeta API Cassandra unavailable"), err, info.Get())
 		}
 		log.Errorf("[GetMeta]\tunknown error\t%+v", err)
-		return nil, status.WrapWithUnknown(fmt.Sprintf("GetMeta API unknown error occurred key %s", key.GetKey()), err, info.Get())
+		return nil, status.WrapWithUnknown(fmt.Sprintf("GetMeta API Cassandra unknown error occurred key %s", key.GetKey()), err, info.Get())
```
Comment (hlts2) [inline pkg/meta/cassandra/handler/grpc/handler.go]: `GetMeta API Cassandra` and `GetMeta API` were mixed in naming, so unify to `GetMeta API Cassandra` .

File: internal/db/nosql/cassandra/cassandra.go [hunk 2]
```diff
@@ -266,18 +267,24 @@ func (c *client) Query(stmt string, names []string) *Queryx {
 	return gocqlx.Query(c.session.Query(stmt), names)
 }
 
+func clientError(err error, keys ...string) error {
+	switch err {
+	case ErrNotFound:
+		return errors.Wrap(err, errors.ErrCassandraNotFound(keys...).Error())
+	case ErrUnavailable:
+		return errors.Wrap(err, errors.ErrCassandraUnavailable().Error())
```
Comment (rinx) [inline internal/db/nosql/cassandra/cassandra.go]: badge
It is better to embed failed keys to error message. it is useful for detecting which keys failed.

File: internal/db/nosql/cassandra/cassandra.go [hunk 3]
```diff
@@ -266,18 +267,24 @@ func (c *client) Query(stmt string, names []string) *Queryx {
 	return gocqlx.Query(c.session.Query(stmt), names)
 }
 
+func clientError(err error, keys ...string) error {
```
Comment (rinx) [inline internal/db/nosql/cassandra/cassandra.go]: badge
the function name does not represent its role.
in this case, it should be "wrapWithKeys", "errorWrappedWithKeys", or "errWithKeys".

File: internal/errors/cassandra.go [hunk 4]
```diff
@@ -30,8 +30,25 @@ var (
...
+	}
+
+	ErrCassandraUnavailable = func() error {
+		return NewErrCassandraUnavailableIdentity()
+	}
+
+	ErrCassandraNotFound = func(keys ...string) error {
+		switch {
+		case len(keys) == 1:
+			return Wrapf(NewErrCassandraNotFoundIdentity(), "error cassandra key '%s' not found", keys[0])
```
Comment (rinx) [inline internal/errors/cassandra.go]: badge
> "error cassandra key '%s' not found"

"error" is useless.
it's better to write just "cassandra key '%s' not found".
It's my mistake.

File: internal/errors/cassandra.go [hunk 5]
```diff
@@ -30,8 +30,25 @@ var (
 		}
 	}
 
-	ErrCassandraNotFound = func(key string) error {
-		return Wrapf(NewErrCassandraNotFoundIdentity(), "error cassandra key '%s' not found", key)
+	NewErrCassandraUnavailableIdentity = func() error {
+		return &ErrCassandraUnavailableIdentity{
+			err: New("error cassandra unavailable"),
```
Comment (rinx) [inline internal/errors/cassandra.go]: same

File: internal/db/nosql/cassandra/cassandra.go [hunk 6]
```diff
@@ -266,18 +267,24 @@ func (c *client) Query(stmt string, names []string) *Queryx {
...
+	switch err {
+	case ErrNotFound:
+		return errors.Wrap(err, errors.ErrCassandraNotFound(keys...).Error())
```
Comment (rinx) [inline internal/db/nosql/cassandra/cassandra.go]: badge
It you wrap the original error with error string of newly generated one, you cannot identify your error in both `IsErrCassandraNotFound` and `IsErrCassandraUnavailable`.
In this case, the original one should be stringify.

File: internal/db/nosql/cassandra/cassandra.go [hunk 7]
```diff
@@ -34,7 +34,8 @@ const (
 )
 
 var (
-	ErrNotFound = gocql.ErrNotFound
+	ErrNotFound    = gocql.ErrNotFound
+	ErrUnavailable = gocql.ErrUnavailable
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/db/nosql/cassandra/cassandra.go]: Thank you for your review.
I wil add thease error.

**Comment by kpango** [comment]
@hlts2 @vankichi how's the progress about this PR?

**Pull request merged**